File: src/Native/GoogleAnalytics.UWP/PlatformInfoProvider.cpp
//
// PlatformInfoProvider.cpp
// Implementation of the PlatformInfoProvider class.
//

#include "pch.h"
#include "PlatformInfoProvider.h"
#include <string>
#include <collection.h>

using namespace GoogleAnalytics;
using namespace Platform;
using namespace Windows::System;
using namespace Windows::UI::Core;
using namespace Windows::UI::ViewManagement;
using namespace Windows::Foundation;
using namespace Windows::Storage;
using namespace Windows::Graphics::Display;


IBox<Dimensions>^ GetDimensions(IBox<Windows::Foundation::Size>^ size)
{
	if (size != nullptr)
	{
		Dimensions result;
		result.Width = size->Value.Width;
		result.Height = size->Value.Height;
		return result;
	}
	return nullptr;
}


IBox<Dimensions>^ ParseDimensionsUpdate(IBox<Dimensions>^currentDimentions,
	IBox<Windows::Foundation::Size>^ newSize,
	bool &hasChanged)
{
	Dimensions valDimensions;
	if (newSize != nullptr)
	{
		valDimensions.Width = newSize->Value.Width;
		valDimensions.Height = newSize->Value.Height;
		if (currentDimentions == nullptr)
		{
			hasChanged = true;
		}
		else if (currentDimentions->Value.Height != newSize->Value.Height ||
			currentDimentions->Value.Width != newSize->Value.Width)
		{
			hasChanged = true;
		}
		return valDimensions;
	}
	else
	{
		if (currentDimentions != nullptr)
		{
			hasChanged = true;
		}
		return nullptr;
	}
}


String^ PlatformInfoProvider::Key_AnonymousClientId = "GoogleAnaltyics.AnonymousClientId";

PlatformInfoProvider::PlatformInfoProvider() :
	windowInitialized(false),
	viewPortResolution(nullptr),
	screenResolution(nullptr)
{
	InitializeWindow();
}

PlatformInfoProvider::~PlatformInfoProvider()
{
	UninitializeWindow();
}

void PlatformInfoProvider::UninitializeWindow()
{
	if (windowInitialized)
	{
		coreWindow->SizeChanged -= sizeChangedEventToken;
		coreWindow = nullptr;
		windowInitialized = false;
	}
}

void PlatformInfoProvider::InitializeWindow()
{
	try
	{
		coreWindow = Windows::UI::Core::CoreWindow::GetForCurrentThread();
		if (coreWindow.Get())
		{
			auto bounds = coreWindow->Bounds;

			float w = bounds.Width;
			float h = bounds.Height;
			auto displayInfo = DisplayInformation::GetForCurrentView();
			switch (displayInfo->ResolutionScale)
			{
			case ResolutionScale::Scale120Percent:
				w = std::floorf(.5f + w * 1.2f);
				w = std::floorf(.5f + w * 1.2f);
				break;
			case ResolutionScale::Scale140Percent:
				w = std::floorf(.5f + w * 1.4f);
				w = std::floorf(.5f + w * 1.4f);
				break;
			case ResolutionScale::Scale150Percent:
				w = std::floorf(.5f + w * 1.5f);
				w = std::floorf(.5f + w * 1.5f);
				break;
			case ResolutionScale::Scale160Percent:
				w = std::floorf(.5f + w * 1.6f);
				w = std::floorf(.5f + w * 1.6f);
				break;
			case ResolutionScale::Scale180Percent:
				w = std::floorf(.5f + w * 1.8f);
				w = std::floorf(.5f + w * 1.8f);
				break;
			case ResolutionScale::Scale225Percent:
				w = std::floorf(.5f + w * 2.25f);
				w = std::floorf(.5f + w * 2.25f);
				break;
			}

			if ((displayInfo->CurrentOrientation & DisplayOrientations::Landscape) == DisplayOrientations::Landscape)
			{
				SetScreenResolution(Size(w, h));
			}
			else // portrait
			{
				SetScreenResolution(Size(h, w));
			}
			SetViewPortResolution(Size(bounds.Width, bounds.Height)); // leave viewport at the scale unadjusted size
			sizeChangedEventToken = coreWindow->SizeChanged += ref new TypedEventHandler<CoreWindow^, WindowSizeChangedEventArgs^>(this, &PlatformInfoProvider::Window_SizeChanged);
			windowInitialized = true;
		}
	}
	catch (const std::exception) { /* ignore, CoreWindow may not be ready yet */ }
}

void PlatformInfoProvider::Window_SizeChanged(CoreWindow^ sender, WindowSizeChangedEventArgs^ e)
{
	SetViewPortResolution(e->Size);
}

void PlatformInfoProvider::OnTracking()
{
	if (!windowInitialized)
	{
		InitializeWindow();
	}
}

String^ PlatformInfoProvider::AnonymousClientId::get()
{
	if (anonymousClientId) return anonymousClientId;
	auto appSettings = ApplicationData::Current->LocalSettings;
	if (!appSettings->Values->HasKey(Key_AnonymousClientId))
	{
		GUID guid;
		CoCreateGuid(&guid);
		std::wstring str(Guid(guid).ToString()->Data());
		auto result = ref new String(str.substr(1, str.length() - 2).data());
		appSettings->Values->Insert(Key_AnonymousClientId, result);
		return result;
	}
	else
	{
		return (String^)appSettings->Values->Lookup(Key_AnonymousClientId);
	}
}

void PlatformInfoProvider::AnonymousClientId::set(String^ value)
{
	anonymousClientId = value;
}

IBox<Dimensions>^ PlatformInfoProvider::ViewPortResolution::get()
{
	return viewPortResolution;
}

void PlatformInfoProvider::SetViewPortResolution(IBox<Size>^ value)
{
	bool hasChanged = false;
	viewPortResolution = ParseDimensionsUpdate(viewPortResolution, value, hasChanged);
	if (hasChanged)
	{
		ViewPortResolutionChanged(this, nullptr);
	}
}

IBox<Dimensions>^ PlatformInfoProvider::ScreenResolution::get()
{
	return screenResolution;
}

void PlatformInfoProvider::SetScreenResolution(IBox<Size>^ value)
{
	bool hasChanged = false;
	screenResolution = ParseDimensionsUpdate(screenResolution, value, hasChanged);
	if (hasChanged)
		ScreenResolutionChanged(this, nullptr);
}

String^ PlatformInfoProvider::UserLanguage::get()
{
	return Windows::Globalization::ApplicationLanguages::Languages->GetAt(0);
}

IBox<int>^ PlatformInfoProvider::ScreenColors::get()
{
	return nullptr;
}

Platform::String^ PlatformInfoProvider::UserAgent::get()
{
	return ConstructUserAgent();
}

String^ PlatformInfoProvider::ConstructUserAgent()
{
	auto sysInfo = ref new Windows::Security::ExchangeActiveSyncProvisioning::EasClientDeviceInformation();
	auto package = Windows::ApplicationModel::Package::Current;
	auto ai = Windows::System::Profile::AnalyticsInfo::VersionInfo;
	String^ sv = ai->DeviceFamilyVersion;

	unsigned long long v = wcstoull(sv->Data(), 0, 10);
	unsigned long long v1 = (v & 0xFFFF000000000000L) >> 48;
	unsigned long long v2 = (v & 0x0000FFFF00000000LL) >> 32;
	String^ systemVersion = v1.ToString() + "." + v2.ToString();

	if (ai->DeviceFamily == "Windows.Desktop") {
		String^ uaArchitecture;
		switch (package->Id->Architecture)
		{
		case ProcessorArchitecture::X64:
			uaArchitecture = "Win64; X64";
			break;
		case ProcessorArchitecture::X86:
			uaArchitecture = "Win32; X86";
			break;
		case ProcessorArchitecture::Arm:
			uaArchitecture = "ARM";
			break;
		default:
			uaArchitecture = "Win64; X64";
			break;
		}
		return "Mozilla/5.0 (Windows NT " + systemVersion + "; " + uaArchitecture + "; " + sysInfo->SystemManufacturer + "; " + sysInfo->SystemProductName + ")";
	}
	else if (ai->DeviceFamily == "Windows.Mobile") {
		return "Mozilla/5.0 (Windows Phone " + systemVersion + "; " + sysInfo->SystemManufacturer + "; " + sysInfo->SystemProductName + ")";
	}
	else if (ai->DeviceFamily == "Windows.Xbox") {
		return "Mozilla/5.0 (Windows NT " + systemVersion + "; Win64; X64; Xbox; " + sysInfo->SystemProductName + ")";
	}
	else {
		return "Mozilla/5.0 (Windows " + systemVersion + "; " + package->Id->Architecture.ToString() + "; " + sysInfo->SystemManufacturer + "; " + sysInfo->SystemProductName + ")";
	}
}

File: src/Managed/GoogleAnalytics.UWP/PlatformInfoProvider.cs
using System;
using System.Linq;
using Windows.ApplicationModel;
using Windows.Devices.Input;
using Windows.Graphics.Display;
using Windows.Storage;
using Windows.System;
using Windows.System.Profile;
using Windows.UI.Core;
using Windows.UI.Xaml;

namespace GoogleAnalytics
{
    /// <summary>
	/// Windows 10, Universal Platform implementation of GoogleAnalytics.Core.IPlatformInfoProvider. 
	/// </summary>
    public sealed class PlatformInfoProvider : IPlatformInfoProvider
    {
        const string Key_AnonymousClientId = "GoogleAnaltyics.AnonymousClientId";

        static string userAgent;

        bool windowInitialized = false;
        string anonymousClientId;
        Dimensions? viewPortResolution;
        Dimensions? screenResolution;
         
        /// <inheritdoc /> 
        public event EventHandler ViewPortResolutionChanged;
        /// <inheritdoc /> 
        public event EventHandler ScreenResolutionChanged;

        public PlatformInfoProvider()
        {
            InitializeWindow();
        }

        /// <inheritdoc /> 
        public void OnTracking()
        {
            if (!windowInitialized)
            {
                InitializeWindow();
            }
        }

        /// <inheritdoc /> 
        public string AnonymousClientId
        {
            get
            {
                if (anonymousClientId == null)
                {
                    var appSettings = ApplicationData.Current.LocalSettings;
                    if (!appSettings.Values.ContainsKey(Key_AnonymousClientId))
                    {
                        anonymousClientId = Guid.NewGuid().ToString();
                        appSettings.Values[Key_AnonymousClientId] = anonymousClientId;
                    }
                    else
                    {
                        anonymousClientId = (string)appSettings.Values[Key_AnonymousClientId];
                    }
                }
                return anonymousClientId;
            }
            set { anonymousClientId = value; }
        }

        /// <inheritdoc /> 
        public Dimensions? ViewPortResolution
        {
            get { return viewPortResolution; }
            private set
            {
                viewPortResolution = value;
                ViewPortResolutionChanged?.Invoke(this, EventArgs.Empty);
            }
        }

        /// <inheritdoc /> 
        public Dimensions? ScreenResolution
        {
            get { return screenResolution; }
            private set
            {
                screenResolution = value;
                ScreenResolutionChanged?.Invoke(this, EventArgs.Empty);
            }
        }

        /// <inheritdoc /> 
        public string UserLanguage
        {
            get { return System.Globalization.CultureInfo.CurrentUICulture.Name; }
        }

        /// <inheritdoc /> 
        /// <remarks>This feature not implemented on this UWP implementation </remarks>
        public int? ScreenColors
        {
            get { return null; }
        }

        /// <inheritdoc /> 
        public string UserAgent
        {
            get
            {
                if (userAgent == null)
                {
                    userAgent = GetUserAgent();
                }
                return userAgent;
            }
        }

        void InitializeWindow()
        {
            try
            {
                if (Window.Current != null && Window.Current.Content != null)
                {
                    var bounds = Window.Current.Bounds;
                    double w, h;
                    var pointerDevice = PointerDevice.GetPointerDevices().FirstOrDefault();
                    if (pointerDevice != null)
                    {
                        w = pointerDevice.ScreenRect.Width;
                        h = pointerDevice.ScreenRect.Height;
                    }
                    else
                    {
                        w = bounds.Width;
                        h = bounds.Height;
                    }

                    var displayInfo = DisplayInformation.GetForCurrentView();
                    var scale = (double)(int)displayInfo.ResolutionScale / 100d;
                    w = Math.Round(w * scale);
                    h = Math.Round(h * scale);

                    if ((displayInfo.NativeOrientation & DisplayOrientations.Landscape) == DisplayOrientations.Landscape)
                    {
                        ScreenResolution = new Dimensions((int)w, (int)h);
                    }
                    else // portrait
                    {
                        ScreenResolution = new Dimensions((int)h, (int)w);
                    }
                    ViewPortResolution = new Dimensions((int)bounds.Width, (int)bounds.Height); // leave viewport at the scale unadjusted size
                    Window.Current.SizeChanged += Current_SizeChanged;
                    windowInitialized = true;
                }
            }
            catch { /* ignore, Bounds may not be ready yet */ }
        }

        void Current_SizeChanged(object sender, WindowSizeChangedEventArgs e)
        {
            ViewPortResolution = new Dimensions((int)e.Size.Width, (int)e.Size.Height);
        }

        static string GetUserAgent()
        {
            var sysInfo = new Windows.Security.ExchangeActiveSyncProvisioning.EasClientDeviceInformation();
            var ai = AnalyticsInfo.VersionInfo;
            string sv = ai.DeviceFamilyVersion;
            ulong v = ulong.Parse(sv);
            ulong v1 = (v & 0xFFFF000000000000L) >> 48;
            ulong v2 = (v & 0x0000FFFF00000000L) >> 32;

            string systemVersion = $"{v1}.{v2}";

            switch (ai.DeviceFamily)
            {
                case "Windows.Desktop":
                    // Edge: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36 Edge/15.15014
                    // This: Mozilla/5.0 (Windows NT 10.0; Win64; X64; Microsoft Corporation; Surface Pro 4)

                    String uaArchitecture;
                    switch (Package.Current.Id.Architecture)
                    {
                        case ProcessorArchitecture.X64:
                            uaArchitecture = "Win64; X64";
                            break;
                        case ProcessorArchitecture.X86:
                            uaArchitecture = "Win32; X86";
                            break;
                        case ProcessorArchitecture.Arm:
                            uaArchitecture = "ARM";
                            break;
                        default:
                            uaArchitecture = "Win64; X64";
                            break;
                    }

                    return $"Mozilla/5.0 (Windows NT {systemVersion}; {uaArchitecture}; {sysInfo.SystemManufacturer}; {sysInfo.SystemProductName})";
                case "Windows.Mobile":
                    // Edge: Mozilla/5.0 (Windows Phone 10.0; Android 6.0.1; Microsoft; Lumia 950) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Mobile Safari/537.36 Edge/15.15014
                    // This: Mozilla/5.0 (Windows Phone 10.0; MicrosoftMDG; RM-1105_12716)

                    return $"Mozilla/5.0 (Windows Phone {systemVersion}; {sysInfo.SystemManufacturer}; {sysInfo.SystemProductName})";
                case "Windows.Xbox":
                    // Edge: Mozilla/5.0 (Windows NT 10.0; Win64; x64; Xbox; Xbox One) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.79 Safari/537.36 Edge/14.14393
                    // This: Mozilla/5.0 (Windows NT 10.0; Win64; X64; Xbox; Xbox One)

                    return $"Mozilla/5.0 (Windows NT {systemVersion}; Win64; X64; Xbox; {sysInfo.SystemProductName})";
                default:
                    return $"Mozilla/5.0 (Windows {systemVersion}; {Package.Current.Id.Architecture.ToString()}; {sysInfo.SystemManufacturer}; {sysInfo.SystemProductName})";
            }
        }
    }
}
